pkg/networkservice/common/connect/server.go

```diff
@@ -65,101 +84,205 @@ func WithDialOptionFactory(factory func(ctx context.Context, request *networkser
+			if request.Connection.Id != "" {
+				// Delete from CE
+				delete(ce.connections, request.Connection.Id)
+				// Delete from global list
+				delete(c.connections, request.Connection.Id)
+			}
+		})
 		return nil, err
 	}
 
+	// Copy conn to request and pass to next one
 	if conn.GetContext() != nil {
 		request.Connection.Context = conn.GetContext()
 	}
+	// we succeed, update connection and remove pending state
+	<-c.executor.AsyncExec(func() {
+		ce.connections[conn.Id] = conn
+
+		// Also update global connection map
+		c.connections[conn.Id] = ce
+	})
+
 	return next.Server(ctx).Request(ctx, request)
 }
 
 func (c *connectServer) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
-	client, found := c.connectionIDToClientMap.Load(conn.GetId())
-	if found {
-		return client.Close(ctx, conn)
+	// Find or Create clientEntry, with add pending.
+	ce, err := c.findOrCreateClient(ctx, conn)
+	if err != nil {
+		// In case of error, no pending state is propogated.
+		return nil, err
 	}
-	u := clienturl.ClientURL(ctx)
-	client, found = c.uRLToClientMap.Load(u.String())
-	if found {
-		c.connectionIDToClientMap.LoadOrStore(conn.GetId(), client)
-		return client.Close(ctx, conn)
+	// Call next Close before calling client Close
+	_, err = next.Server(ctx).Close(ctx, conn)
+
+	// Call client close
+	if _, closeErr := ce.client.Close(ctx, conn); closeErr != nil {
+		if err != nil {
+			// Combine errors
+			err = errors.Wrapf(err, "errors during close: %v", closeErr)
+		}
 	}
-	// TODO - do something with cancelFunc to clean up uRLToClientMap when we are done with them (maybe ref counting on Close?)
+
+	// remove connection from list
+	<-c.executor.AsyncExec(func() {
+		delete(ce.connections, conn.Id)
+		delete(c.connections, conn.Id)
+		// Close client if there is no more users for it.
+		c.closeClient(ctx, ce)
+	})
+
+	return &empty.Empty{}, err
+}
+
+func (c *connectServer) findOrCreateClient(ctx context.Context, conn *networkservice.Connection) (ce *clientEntry, err error) {
+	connectionExists := false
+	clientURI := clienturl.ClientURL(ctx)
+	<-c.executor.AsyncExec(func() {
+		if clientURI != nil {
+			ce, connectionExists = c.clientMap[clientURI.String()]
+		}
+		if !connectionExists {
+			// try to find using connection id
+			ce, connectionExists = c.connections[conn.Id]
+		}
+		if !connectionExists {
+			// Since there is no connection, we need to create it and mark as pending as well
+			// We could also be sure we one who will open connection.
+			if clientURI == nil {
+				ce.err = errors.Errorf("failed to create connection %v context should contain clienturl.ClientURL", conn)
+				return
+			}
+			ce = &clientEntry{
+				clientURI:   clientURI,
+				ready:       make(chan bool, 1),
+				connections: map[string]*networkservice.Connection{},
+			}
+			// Put/Update connection
+			c.clientMap[clientURI.String()] = ce
+		}
+		ce.connections[conn.Id] = conn
+	})
+	if ce.err != nil {
```

Comment:
I don't quite follow why we are doing this here rather than at the point we are assigning non-nil values to c.err

Reply:
Simplified, but it will require if clientURI is not passed to check for connections for it.